dotnet/WPF/VisualLayerIntegration/VisualLayerIntegration/BarGraphHostControl.xaml.cs

```diff
@@ -0,0 +1,197 @@
+        {
+            // Adjust light position.
+            if (currentGraph != null)
+            {
+                // Convert mouse position to DIP (is raised in physical pixels).
+                var posDip = GetPointInDIP(e.point);
+
+                var adjustedTopLeft = GetControlPointInDIP(CompositionHostElement);
+
+                // Get point relative to control.
+                var relativePoint = new SysWin.Point(posDip.X - adjustedTopLeft.X, posDip.Y - adjustedTopLeft.Y);
+
+                // Update light position.
+                currentGraph.UpdateLight(relativePoint);
+            }
+        }
+
+        private SysWin.Point GetPointInDIP(SysWin.Point point)
+        {
+            var posDipX = point.X / (currentDpiX / 96.0);
+            var posDipY = point.Y / (currentDpiY / 96.0);
+            return new SysWin.Point(posDipX, posDipY);
+        }
+
+        private SysWin.Point GetControlPointInDIP(UIElement control)
+        {
+            // Get bounds of hwnd host control.
+            // Top left of control relative to screen.
+            var controlTopLeft = control.PointToScreen(new SysWin.Point(0, 0));
+            // Convert screen coord to DIP.
+            var adjustedX = controlTopLeft.X / (currentDpiX / 96.0);
+            var adjustedY = controlTopLeft.Y / (currentDpiY / 96.0);
+            return new SysWin.Point(adjustedX, adjustedY);
+        }
+
+        protected override void OnDpiChanged(DpiScale oldDpi, DpiScale newDpi)
+        {
+            if (this.ActualWidth > 0 && currentGraph != null)
+            {
+                currentGraph.UpdateSize(newDpi, CompositionHostElement.ActualWidth, CompositionHostElement.ActualHeight);
+            }
+        }
+
+        // Handle Composition tree creation and updates.
+        public void UpdateGraph()
+        {
+            Customer customer = DataContext as Customer;
+            if (customer != null)
+            {
+                var graphTitle = customer.FirstName + " Investment History";
+                var xAxisTitle = "Investment #";
+                var yAxisTitle = "# Shares of Stock";
+
+                // If graph already exists update values. Otherwise, create new graph.
+                if (graphContainer.Children.Count > 0 && currentGraph != null)
+                {
+                    currentGraph.UpdateGraphData(graphTitle, xAxisTitle, yAxisTitle, customer.Data);
+                }
+                else
+                {
+                    var graph = new BarGraph(compositor, compositionHost.hwndHost, graphTitle, xAxisTitle, yAxisTitle,
+                        (float)CompositionHostElement.ActualWidth, (float)CompositionHostElement.ActualHeight, currentDpiX, currentDpiY, customer.Data, windowRenderTarget,
+                        true, BarGraph.GraphBarStyle.PerBarLinearGradient,
+                        new List<Windows.UI.Color> { Windows.UI.Color.FromArgb(255, 246, 65, 108), Windows.UI.Color.FromArgb(255, 255, 246, 183) });
```

Comment:
Do these colors have names you could use (to save the reader having to guess what colors they are)?

Reply:
not that I know of, these are custom